## Part 1: Explore the Data

Import the data and use Pandas to learn more about the dataset.

In [3]:
import pandas as pd

df = pd.read_csv('Resources/client_dataset.csv')

df.head()

first      last                    job               phone  \
0     Donald   Harding           Immunologist  793-904-7725x39308   
1    Tiffany     Myers        Music therapist    201.442.4543x942   
2    Shannon    Watson           Immunologist   687.737.9424x8503   
3     Nathan     Baker  Accounting technician    827-788-8123x012   
4  Christina  Schwartz           Chiropractor        265-829-3643   

                                 email  client_id  order_id  order_date  \
0     harding.donald.7185@sullivan.com      58515   8953482  2023-04-28   
1    myers.t.6537@ferguson-johnson.net      37609   8069089  2023-05-19   
2                swatson8146@payne.net      57113   1902144  2023-01-29   
3               bakernathan@benson.com      46554   9031802  2023-04-25   
4  christinaschwartz9252@mcconnell.com      92089   1322274  2023-05-28   

   order_week  order_year         item_id     category         subcategory  \
0          17        2023  EUD29711-63-6U        decor            wall art   
1          20        2023  XDA18116-89-4A  consumables                pens   
2           4        2023  ABE59463-05-7E     software  project management   
3          17        2023  ZMM00836-65-0C  consumables                pens   
4          21        2023  BZX55559-12-3X  consumables                misc   

   unit_price  unit_cost  unit_weight  qty  line_number  
0     1096.80     762.71         7.50  105            1  
1       24.95      15.09         1.49   21            0  
2       13.52       7.86         1.68   39            6  
3       36.42      24.85         1.23   29            3  
4      195.10     108.17        46.43   20            1

In [4]:
# View the column names in the data
df.columns 



Index(['first', 'last', 'job', 'phone', 'email', 'client_id', 'order_id',
       'order_date', 'order_week', 'order_year', 'item_id', 'category',
       'subcategory', 'unit_price', 'unit_cost', 'unit_weight', 'qty',
       'line_number'],
      dtype='object')

In [5]:
# Use the describe function to gather some basic statistics
df.describe ()


client_id      order_id    order_week    order_year    unit_price  \
count  54639.000000  5.463900e+04  54639.000000  54639.000000  54639.000000   
mean   54837.869416  5.470190e+06     11.359139   2022.993064    136.267207   
std    25487.438231  2.599807e+06      7.023499      0.082997    183.873135   
min    10033.000000  1.000886e+06      1.000000   2022.000000      0.010000   
25%    33593.000000  3.196372e+06      6.000000   2023.000000     20.800000   
50%    53305.000000  5.496966e+06     11.000000   2023.000000     68.310000   
75%    78498.000000  7.733869e+06     17.000000   2023.000000    173.160000   
max    99984.000000  9.998480e+06     52.000000   2023.000000   1396.230000   

          unit_cost   unit_weight           qty   line_number  
count  54639.000000  54639.000000  5.463900e+04  54639.000000  
mean      99.446073      5.004116  5.702646e+02      2.979667  
std      133.164267      5.326599  1.879552e+04      2.436320  
min        0.010000      0.000000  0.000000e+00      0.000000  
25%       14.840000      1.450000  3.200000e+01      1.000000  
50%       49.890000      3.240000  6.800000e+01      3.000000  
75%      125.570000      6.890000  1.700000e+02      5.000000  
max      846.270000     46.430000  3.958244e+06      9.000000

In [90]:
# Use this space to do any additional research
# and familiarize yourself with the data.



In [93]:
# What three item categories had the most entries?
df["category"].value_counts().head(3)

category
consumables    23538
furniture      11915
software        8400
Name: count, dtype: int64

In [99]:
# For the category with the most entries,
# which subcategory had the most entries?
new_category= df.loc[df["category"]=="consumables", "subcategory"].value_counts().head(5)
new_category

subcategory
bathroom supplies    6424
paper products       5874
kitchen supplies     5746
misc                 2891
pens                 2603
Name: count, dtype: int64

In [ ]:
df.loc[df["category"]=="consumables", "subcategory"].value_counts().head(5)

In [97]:
# Which five clients had the most entries in the data?
five_clients = df['client_id'].value_counts()
five_clients.head(5)


client_id
33615    220
66037    211
46820    209
38378    207
24741    207
Name: count, dtype: int64

In [30]:
# Store the client ids of those top 5 clients in a list.
list_five_clients= list(df['client_id'].value_counts().head(5).index)
list_five_clients

[33615, 66037, 46820, 38378, 24741]

In [31]:
# How many total units (the qty column) did the
# client with the most entries order order?
df.loc[df["client_id"]== 33615, "qty"]. sum()


64313

## Part 2: Transform the Data
Do we know that this client spent the more money than client 66037? If not, how would we find out? Transform the data using the steps below to prepare it for analysis.



In [32]:
# Create a column that calculates the 
# subtotal for each line using the unit_price
# and the qty
df["subtotal of unit price* qty"] = df["unit_price"] * df["qty"]
df.head()


first      last                    job               phone  \
0     Donald   Harding           Immunologist  793-904-7725x39308   
1    Tiffany     Myers        Music therapist    201.442.4543x942   
2    Shannon    Watson           Immunologist   687.737.9424x8503   
3     Nathan     Baker  Accounting technician    827-788-8123x012   
4  Christina  Schwartz           Chiropractor        265-829-3643   

                                 email  client_id  order_id  order_date  \
0     harding.donald.7185@sullivan.com      58515   8953482  2023-04-28   
1    myers.t.6537@ferguson-johnson.net      37609   8069089  2023-05-19   
2                swatson8146@payne.net      57113   1902144  2023-01-29   
3               bakernathan@benson.com      46554   9031802  2023-04-25   
4  christinaschwartz9252@mcconnell.com      92089   1322274  2023-05-28   

   order_week  order_year  ...         subcategory unit_price unit_cost  \
0          17        2023  ...            wall art    1096.80    762.71   
1          20        2023  ...                pens      24.95     15.09   
2           4        2023  ...  project management      13.52      7.86   
3          17        2023  ...                pens      36.42     24.85   
4          21        2023  ...                misc     195.10    108.17   

   unit_weight  qty  line_number   subtotal  subtotal of unit price* qty  \
0         7.50  105            1  115164.00                    115164.00   
1         1.49   21            0     523.95                       523.95   
2         1.68   39            6     527.28                       527.28   
3         1.23   29            3    1056.18                      1056.18   
4        46.43   20            1    3902.00                      3902.00   

   total_weight  shipping_price  
0        787.50         5512.50  
1         31.29          312.90  
2         65.52          458.64  
3         35.67          356.70  
4        928.60         6500.20  

[5 rows x 22 columns]

In [37]:
# Create a column for shipping price.
# Assume a shipping price of $7 per pound
# for orders over 50 pounds and $10 per
# pound for items 50 pounds or under.
def shipping_price_calc(weight):
    if weight > 50:
        return weight * 7
    return weight * 10 

df['total_weight'] = df['qty'] * df['unit_weight']
df['shipping_price'] = df['total_weight'].apply(shipping_price_calc)
df[['unit_price', 'unit_weight', 'qty', 'total_weight', 'shipping_price']].head(3)

unit_price  unit_weight  qty  total_weight  shipping_price
0     1096.80         7.50  105        787.50         5512.50
1       24.95         1.49   21         31.29          312.90
2       13.52         1.68   39         65.52          458.64

In [39]:
# Create a column for the total price
# using the subtotal and the shipping price
# along with a sales tax of 9.25%

df['line_price'] = round((df['line_subtotal'] + df['shipping_price']) * 1.0925, 2)
df[['line_subtotal', 'shipping_price', 'line_price']].head(3)

KeyError: 'line_subtotal'

In [23]:
# Create a column for the cost
# of each line using unit cost, qty, and
# shipping price (assume the shipping cost
# is exactly what is charged to the client).

df['line_cost'] = df['unit_cost'] * df['qty'] + df['shipping_price']
df.head(3)

KeyError: 'shipping_price'

In [40]:
# Create a column for the profit of
# each line using line cost and line price

df['line_profit'] = df['line_price'] - df['line_cost']
df.head(3)

KeyError: 'line_price'

## Part 3: Confirm your work
You have email receipts showing that the total prices for 3 orders. Confirm that your calculations match the receipts. Remember, each order has multiple lines.

Order ID 2742071 had a total price of \$152,811.89

Order ID 2173913 had a total price of \$162,388.71

Order ID 6128929 had a total price of \$923,441.25


In [41]:
# Check your work using the totals above
order_ids = [2742071, 2173913, 6128929]

def find_order_price(order_id):
    order_lines_df = df.loc[df['order_id'] == order_id, 'line_price']
    return round(order_lines_df.sum(), 2)

for order_id in order_ids:
    print(f"Order {order_id} total: ${find_order_price(order_id)}")

KeyError: 'line_price'

## Part 4: Summarize and Analyze
Use the new columns with confirmed values to find the following information.

In [15]:
# How much did each of the top 5 clients by quantity
# spend? Check your work from Part 1 for client ids.



In [16]:
# Create a summary DataFrame showing the totals for the
# for the top 5 clients with the following information:
# total units purchased, total shipping price,
# total revenue, and total profit. Sort by total profit.



In [17]:
# Format the data and rename the columns
# to names suitable for presentation.
# Currency should be in millions of dollars.



In [18]:
# Sort the updated data by "Total Profit" form highest to lowest
